### PR TITLE
feat(1319): build cluster queue worker / consumer

### PR DESCRIPTION
**Pull request by parthasl**
Context:
Multiple build cluster 

Objective:
Consume messages from build cluster queue and submit to k8s

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -0,0 +1,5231 @@
+{
```
Comment (d2lam) [inline package-lock.json]: can remove this file, it adds a lot of noise

Comment (parthasl) [reply package-lock.json]: removed, but i see this in most of the repos

Comment (d2lam) [reply package-lock.json]: 👍 it was automatically added at one point when node is upgraded. But we decided to remove afterward.

File: test/.eslintrc.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+extends: screwdriver
```
Comment (tkyi) [inline test/.eslintrc.yaml]: what is this file for

File: test/helper.test.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+'use strict';
+
+const assert = require('chai').assert;
+const mockery = require('mockery');
+const sinon = require('sinon');
+
+sinon.assert.expose(assert, { prefix: '' });
+
+describe('Helper Test', () => {
+    const fullBuildConfig = JSON.parse('{"job":"stop",' +
```
Comment (tkyi) [inline test/helper.test.js]: it's probably easier to read/handle if you put this in a separate json file
eg: https://github.com/screwdriver-cd/config-parser/blob/master/test/data/bad-build-cluster.json
https://github.com/screwdriver-cd/config-parser/blob/master/test/index.test.js#L183

File: index.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+const logger = winston.createLogger({
+    transports: [
+        new winston.transports.Console()
+    ]
+});
+
+let channelWrapper;
+
+/**
+ * [onMessage consume messages in batches, once its available in the queue. channelWrapper has in-built back pressure
```
Comment (tkyi) [inline index.js]: don't need brackets around any of the descriptions, either

File: lib/helper.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+'use strict';
+
+const request = require('request');
+
+/**
+ * [updateBuildStatus with failure reason]
+ * @param  {Object}     buildConfig   [build config of the job]
+ * @param  {[String]}   status        [FAILED]
```
Comment (tkyi) [inline lib/helper.js]: Extra brackets and missing param type for statusMessage.

File: test/.eslintrc.yaml (hunk 2 above)
Comment (tkyi) [reply test/.eslintrc.yaml]: can you change line #41 from `mockExecutorRouter = function () { return mockExecutor; };` to `mockExecutorRouter = () => mockExecutor;`?

File: lib/jobs.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    ecosystem
+});
+const logger = winston.createLogger({
+    transports: [
+        new winston.transports.Console()
+    ]
+});
+
+/**
+ * [k8sExectutor start or stop build]
```
Comment (tkyi) [inline lib/jobs.js]: You can see an example of how we usually do jsdocs here:
https://github.com/screwdriver-cd/models/blob/master/lib/build.js#L28-L35

Don't need all the brackets `[]` around everything.

File: test/.eslintrc.yaml (hunk 2 above)
Comment (parthasl) [reply test/.eslintrc.yaml]: its not working, i just followed below repo.
refer this -> https://github.com/screwdriver-cd/queue-worker/blob/master/test/jobs.test.js#L86
https://github.com/screwdriver-cd/queue-worker/tree/master/test

File: lib/helper.js (hunk 5 above)
Comment (parthasl) [reply lib/helper.js]: done

File: lib/jobs.js (hunk 6 above)
Comment (parthasl) [reply lib/jobs.js]: fixed

File: test/.eslintrc.yaml (hunk 2 above)
Comment (tkyi) [reply test/.eslintrc.yaml]: ah okay

File: index.js [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+let channelWrapper;
+
+/**
+ * onMessage consume messages in batches, once its available in the queue. channelWrapper has in-built back pressure
+ *            meaning if consumed messages are not ack'd or nack'd, it will not fetch more messages. Definitely need
+ *            to ack or nack messages, otherwise it will halt indefinitely. submit start or stop jobs to k8s executor
+ *            using threads
+ * @param  {Object} data  Message from queue with headers, timestamp, and other properties; will be used to ack or nack the message
+ * @return {}             none
```
Comment (tkyi) [inline index.js]: can delete this line

File: lib/jobs.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+});
+const logger = winston.createLogger({
+    transports: [
+        new winston.transports.Console()
+    ]
+});
+
+/**
+ * k8sExectutor start or stop build
+ * @param  {String}   jobType start or stop
```
Comment (tkyi) [inline lib/jobs.js]: missing `buildConfig` and `job`

File: lib/jobs.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+ * @return {promise}          return promise
+ */
+module.exports = function k8sExectutor([jobType, buildConfig, job]) {
```
Comment (tkyi) [inline lib/jobs.js]: k8sExec~t~utor
extra `t` in the middle

File: lib/jobs.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+});
+
+/**
+ * k8sExecutor start or stop build
+ * @param  {String}   jobType     start or stop
+ * @param  {Object}   buildConfig build config
+ * @param  {String}   job         concatenated log message jobId, jobType, buildId
+ * @return {promise}              return promise
+ */
+module.exports = function k8sExecutor([jobType, buildConfig, job]) {
```
[the remark this answers is not in the document]
Comment (parthasl) [reply lib/jobs.js]: my assumption is it's going to be in k8s. changed k8sExecutor to buildExecutor

**Pull request merged**